Add support for getting a joypad's battery and connection information

Nintorch opened the pull request:
- Closes https://github.com/godotengine/godot-proposals/issues/14482

I decided to split my big SDL3 joypad features PR ( https://github.com/godotengine/godot/pull/107967 ) into several smaller PRs, this PR is one of them.

This PR adds the ability to get information about the joypad's battery state (no battery, charging, fully charged), battery percentage, connection state (wired, wireless).

Relevant SDL documentation:
[SDL_GetJoystickPowerInfo](https://wiki.libsdl.org/SDL3/SDL_GetJoystickPowerInfo)
[SDL_GetJoystickConnectionState](https://wiki.libsdl.org/SDL3/SDL_GetJoystickConnectionState)

This PR used to have a name `Add support for querying more joypad information`, but was repurposed to `Add support for getting a joypad's battery and connection information` since getting button/axis names and `has_joy_axis/button()` might not be necessary to be builtin in Godot, and `get_joy_model()`, `get_joy_type()` have been moved into a separate PR ( https://github.com/godotengine/godot/pull/117526 ).
<details>
<summary>Old TODO</summary>

- [ ] `JOY_MODEL_STEAM_DECK` documentation
- ~~Throttle documentation~~ I don't have access to one :(
- ~~`get_joy_num_axes/buttons()`~~ I think checking if a joypad has a particular axis/button is more useful
- [x] `has_joy_axis/button()`
- [x] Mark as experimental
- [x] Supersede https://github.com/godotengine/godot/pull/89193
- ~~`get_joy_model()` and `get_joy_scheme()` for Android and Web (depends on https://github.com/godotengine/godot/pull/114338 and https://github.com/godotengine/godot/pull/114589 )~~ (Support for other platforms will be done in a separate PR)
- ~~`get_joy_power_state()` and `get_joy_battery_percent()` for Android ( https://stackoverflow.com/questions/78331095/how-to-get-battery-level-of-a-game-controller-from-an-android-application and https://developer.android.com/reference/android/hardware/BatteryState )~~
- ~~`has_joy_axis/button()` for Android and Web based on Godot's mapping system~~
- [x] Mark as available on iOS (see also https://github.com/godotengine/godot/pull/114316 )
- [x] Remove button/axis name functionality
- [x] Remove `get_joy_model()` and `get_joy_type()` functionality since it's been moved into a separate PR ( https://github.com/godotengine/godot/pull/117526 )
</details>

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		</constant>
+		<constant name="JOY_MODEL_XBOX_GENERIC" value="0" enum="JoyModel">
+			A standard game controller model with the button layout similar to Xbox controllers.
+			A game controller with this model ID has A/B/X/Y face buttons, a directional pad, and it might have Back/Guide/Start buttons, left/right shoulder buttons and triggers.
```
```suggestion
			A game controller with this model ID has A/B/X/Y face buttons, a directional pad, and it might have Back/Guide/Start buttons, left/right shoulder buttons, and triggers.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		</constant>
+		<constant name="JOY_MODEL_PS3" value="6" enum="JoyModel">
+			A game controller model that represents the PlayStation 3 (Dualshock 3) controller.
+			A game controller with this model ID has Cross/Circle/Square/Triangle face buttons, a directional pad, left and right sticks, Select/PS/Start buttons, L1/R1 shoulder buttons, L2/R2 trigger axes, accelerometer and gyroscope sensors.
```
```suggestion
			A game controller with this model ID has Cross/Circle/Square/Triangle face buttons, a directional pad, left and right sticks, Select/PS/Start buttons, L1/R1 shoulder buttons, L2/R2 trigger axes, and accelerometer and gyroscope sensors.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		</constant>
+		<constant name="JOY_MODEL_SWITCH_PRO" value="9" enum="JoyModel">
+			A game controller model that represents the Nintendo Switch Pro Controller.
+			A game controller with this model ID has B/A/Y/X face buttons, a directional pad, left and right sticks, Minus/Home/Plus buttons, L/R shoulder buttons, ZL/ZR trigger axes (with a value of either 0 or 1), a Capture button, accelerometer and gyroscope sensors.
```
```suggestion
			A game controller with this model ID has B/A/Y/X face buttons, a directional pad, left and right sticks, Minus/Home/Plus buttons, L/R shoulder buttons, ZL/ZR trigger axes (with a value of either [code]0[/code] or [code]1[/code]), a Capture button, and accelerometer and gyroscope sensors.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		</constant>
+		<constant name="JOY_MODEL_JOYCON_LEFT" value="10" enum="JoyModel">
+			A game controller model that represents the left Nintendo Switch Joycon that is held horizontally.
+			A game controller with this model ID has Face South/Face East/Face West/Face North face buttons, a left stick, SL and SR buttons, Minus/Capture/L/ZL buttons, accelerometer and gyroscope sensors.
```
```suggestion
			A game controller with this model ID has Face South/Face East/Face West/Face North face buttons, a left stick, SL and SR buttons, Minus/Capture/L/ZL buttons, and accelerometer and gyroscope sensors.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		<constant name="JOY_SCHEME_PLAYSTATION" value="1" enum="JoyScheme">
+			A game controller layout scheme that represents the PlayStation controller layout.
+			A game controller with this layout scheme has Cross/Circle/Square/Triangle face buttons,
+			a directional pad, and it may have left and right sticks, Select/PS/Start (or Share/PS/Options) buttons, L1/R1 shoulder buttons, L2/R2 trigger axes and a touchpad in case of PS4 and PS5 controllers.
```
```suggestion
			A game controller with this layout scheme has Cross/Circle/Square/Triangle face buttons, a directional pad, and it may have left and right sticks, Select/PS/Start (or Share/PS/Options) buttons, L1/R1 shoulder buttons, L2/R2 trigger axes, and a touchpad in the case of PS4 and PS5 controllers.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		</constant>
+		<constant name="JOY_SCHEME_NINTENDO" value="2" enum="JoyScheme">
+			A game controller layout scheme that represents the Nintendo controller layout.
+			A game controller with this layout scheme has B/A/Y/X face buttons, a directional pad, and it may have left and right sticks, Minus/Home/Plus (or Start/Home/Select) buttons, L/R shoulder buttons, ZL/ZR trigger axes, and possibly a Capture button in case of Nintendo Switch Pro Controller.
```
```suggestion
			A game controller with this layout scheme has B/A/Y/X face buttons, a directional pad, and it may have left and right sticks, Minus/Home/Plus (or Start/Home/Select) buttons, L/R shoulder buttons, ZL/ZR trigger axes, and possibly a Capture button in the case of Nintendo Switch Pro Controller.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		</constant>
+		<constant name="JOY_SCHEME_JOYCON_HORIZONTAL" value="3" enum="JoyScheme">
+			A game controller layout scheme that represents the layout of a Nintendo Switch Joycon controller that is being held horizontally.
+			A game controller with this layout scheme has Face South/Face East/Face West/Face North face buttons, a left stick, SL and SR buttons and either Minus/Capture/L/ZL or Plus/Home/R/ZR buttons.
```
```suggestion
			A game controller with this layout scheme has Face South/Face East/Face West/Face North face buttons, a left stick, SL and SR buttons, and either Minus/Capture/L/ZL or Plus/Home/R/ZR buttons.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+			Includes a pitch and roll stick that reports as the Left Stick, a POV Hat which reports as the Right Stick,
+			a rudder (handle twist or rocker) that reports as Left Trigger, and a throttle control as the Right Trigger.
+			Includes support for a primary weapon (A), secondary weapon (B), and other standard buttons (X, Y, START, BACK).
+			LB, LSB, RB, and RSB are optional.
```
```suggestion
			Includes a pitch and roll stick that reports as the Left Stick, a POV Hat which reports as the Right Stick, a rudder (handle twist or rocker) that reports as Left Trigger, and a throttle control as the Right Trigger.
			Includes support for a primary weapon (A), secondary weapon (B), and other standard buttons (X, Y, START, BACK). LB, LSB, RB, and RSB are optional.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+		<constant name="JOY_DEVICE_GUITAR" value="6" enum="JoyDeviceType">
+			Guitar controller.
+			The strum bar maps to DPAD (up and down), and the frets are assigned to A (green), B (red), Y (yellow), X (blue), and LB (orange).
+			Right Stick Y is associated with a vertical orientation sensor; Right Stick X is the whammy bar.
```
```suggestion
			The Right Stick Y is associated with a vertical orientation sensor; the Right Stick X is the whammy bar.
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,155 @@
+			Arcade pad controller.
+			Includes Directional Pad and most standard buttons (A, B, X, Y, START, BACK, LB, RB).
+			The Left and Right Triggers are implemented as digital buttons and report either 0 or 1.
+			Left Stick, Right Stick, LSB, and RSB are optional.
```
```suggestion
			The Left and Right Triggers are implemented as digital buttons and report either [code]0[/code] or [code]1[/code]. Left Stick, Right Stick, LSB, and RSB are optional.
```

Nintorch commented on core/core_constants.cpp:
```diff
@@ -562,6 +562,48 @@ void register_global_constants() {
+	BIND_CORE_ENUM_CLASS_CONSTANT(JoyModel, JOY_MODEL, JOYCON_LEFT);
+	BIND_CORE_ENUM_CLASS_CONSTANT(JoyModel, JOY_MODEL, JOYCON_RIGHT);
+	BIND_CORE_ENUM_CLASS_CONSTANT(JoyModel, JOY_MODEL, JOYCON_PAIR);
+	BIND_CORE_ENUM_CLASS_CONSTANT(JoyModel, JOY_MODEL, STEAM_DECK);
```
I will need to move this line so it matches the order in the enum

Nintorch commented on core/input/input.h:
```diff
@@ -182,6 +182,11 @@ class Input : public Object {
 		int hat_current = 0;
+		JoyModel model = JoyModel::INVALID;
+		JoyDeviceType device_type = JoyDeviceType::UNKNOWN;
+		int battery_percent = -1;
```
It's probably better to store it as `uint8_t`

Nintorch commented on core/input/input_enums.h:
```diff
@@ -104,6 +104,62 @@ enum class JoyButton {
 
+// See SDL_GamepadType
+enum class JoyModel {
+	INVALID = -1,
```
```suggestion
	UNKNOWN = 0,
```
For consistency

Nintorch commented on doc/classes/Input.xml:
```diff
@@ -140,13 +164,37 @@
+			<return type="int" enum="JoyScheme" />
+			<param index="0" name="device" type="int" />
+			<description>
+				Returns the joypad's scheme layout (i.e. if it's an Xbox controller, a PlayStation controller, a Nintendo controller, etc.).
```
I need to mention that this method should mostly be used to determine the face buttons layout

Nintorch commented on core/input/input.h:
```diff
@@ -182,6 +182,11 @@ class Input : public Object {
 		int hat_current = 0;
+		JoyModel model = JoyModel::UNKNOWN;
+		JoyDeviceType device_type = JoyDeviceType::UNKNOWN;
+		uint8_t battery_percent = -1;
```
`uint8_t` can't store signed values, I should change it to `int8_t`

Nintorch commented on core/input/input_enums.h:
```diff
@@ -104,6 +104,64 @@ enum class JoyButton {
+	FULL_BATTERY = 4,
+};
+
+// See SDL_JoystickConnectionState
```
```suggestion
// See SDL_JoystickConnectionState. SDL_JOYSTICK_CONNECTION_INVALID equivalent was omitted for simplicity
```

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,149 @@
+		</constant>
+		<constant name="JOY_SCHEME_PLAYSTATION" value="2" enum="JoyScheme">
+			A game controller layout scheme that represents the PlayStation controller layout.
+			A game controller with this layout scheme has Cross/Circle/Square/Triangle face buttons, a directional pad, and it may have left and right sticks, Select/PS/Start (or Share/PS/Options) buttons, L1/R1 shoulder buttons, L2/R2 trigger axes, and a touchpad in the case of PS4 and PS5 controllers.
```
```suggestion
			A game controller with this layout scheme has Cross/Circle/Square/Triangle face buttons, a directional pad, and it may have left and right sticks, Select/PS/Start (or Share/PS/Options) buttons, L1/R1 shoulder buttons, L2/R2 trigger axes, and a touchpad in the case of the PS4 and PS5 controllers.
```

Nintorch commented:
Sure, I can prepare a test project later!

DanielSnd commented:
> Sure, I can prepare a test project later!

Sounds good then :)

Nintorch commented:
Here :D
[joypad-info-test.zip](https://github.com/user-attachments/files/24414989/joypad-info-test.zip)

And I just found a web page with Steam Deck's buttons and their names, so we will use it to get more specific button names.

May I ask if you can test the project above, verify that the model name is "Steam Deck", press every button on the Steam Deck that is usable in games and save the reported names for them that are shown in "Last pressed button" label?

Nintorch commented:
I might move the buttons/axes names into a separate PR, I think I have a better way to make this functionality
I thought about something like `Input.get_action_input_names(device = -1)` that allows the developer to retrieve the button/axis names (not IDs) from input actions, and the names will depend on the joypad's model and scheme.
Basically, if there are keyboard buttons and joypad buttons in an action, this method would return keyboard button names if the last used device wasn't a joypad, and joypad button names otherwise.
Getting button IDs makes sense, but they're separate enums, so multiple methods would have to be created, and I think `InputMap.action_get_events()` (I hope I got the name right) is better suited for this.

AThousandShips commented on doc/classes/@GlobalScope.xml:
```diff
@@ -2550,6 +2550,149 @@
+		</constant>
+		<constant name="JOY_MODEL_JOYCON_PAIR" value="13" enum="JoyModel">
+			A game controller model that represents the pair of Nintendo Switch Joycon controllers, with a similar layout to the Nintendo Switch Pro Controller.
+			A game controller with this model ID has B/A/Y/X face buttons, a directional pad, left and right sticks, Minus/Home/Plus buttons, L/R shoulder buttons, ZL/ZR trigger axes (with a value of either [code]0[/code] or [code]1[/code]), a Capture button, and accelerometer and gyroscope sensors (the sensors of the right Joycon controller are being used).
```
```suggestion
			A game controller with this model ID has B/A/Y/X face buttons, a directional pad, left and right sticks, Minus/Home/Plus buttons, L/R shoulder buttons, ZL/ZR trigger axes (with a value of either [code]0[/code] or [code]1[/code]), a Capture button, and accelerometer and gyroscope sensors (the sensors of the right Joycon controller is used).
```

Nintorch committed:
Add support for joypad battery and connection info

This PR adds the ability to get information about the joypad's battery state (no battery, charging, fully charged), battery percentage, connection state (wired, wireless).
